modules/carousel/jetpack-carousel.php

```diff
@@ -540,9 +540,57 @@ class_exists( 'Jetpack_AMP_Support' )
 			 * @param array $extra_data Array of data about the site and the post.
 			 */
 			$extra_data = apply_filters( 'jp_carousel_add_data_to_container', $extra_data );
+
 			foreach ( (array) $extra_data as $data_key => $data_values ) {
-				$html = str_replace( '<div ', '<div ' . esc_attr( $data_key ) . "='" . json_encode( $data_values ) . "' ", $html );
-				$html = str_replace( '<ul class="wp-block-gallery', '<ul ' . esc_attr( $data_key ) . "='" . json_encode( $data_values ) . "' class=\"wp-block-gallery", $html );
+				// Do not go any further if DOMDocument is disabled on the server.
+				if ( ! class_exists( 'DOMDocument' ) ) {
+					return $html;
+				}
+
+				// Let's grab all containers from the HTML.
+				$dom_doc = new DOMDocument();
+
+				/*
+				 * The @ is not enough to suppress errors when dealing with libxml,
+				 * we have to tell it directly how we want to handle errors.
+				 */
+				libxml_use_internal_errors( true );
+				@$dom_doc->loadHTML( $html ); // phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged
+				libxml_use_internal_errors( false );
+
+				// Let's look for lists and divs.
+				$ul_tags  = $dom_doc->getElementsByTagName( 'ul' );
+				$div_tags = $dom_doc->getElementsByTagName( 'div' );
+
+				// Loop through each ul, and add the data to it if it is a gallery block.
+				foreach ( $ul_tags as $ul_tag ) {
+					if ( false !== strpos( $ul_tag->getAttribute( 'class' ), 'wp-block-gallery' ) ) {
+						$ul_tag->setAttribute(
+							esc_attr( $data_key ),
```

Comment:
We should not need to `esc_attr()` here.
```suggestion
						$data_key,
```